File: go/ink/encryptionpolicy_test.go
package ink

import (
	"sort"
	"testing"
)

func TestConfidentialIntentsMatchTheSpecSet(t *testing.T) {
	got := append([]string(nil), ConfidentialIntents...)
	sort.Strings(got)
	want := []string{"context_share", "multi_party_sync", "schedule_meeting"}
	if len(got) != len(want) {
		t.Fatalf("got %v", got)
	}
	for i := range want {
		if got[i] != want[i] {
			t.Fatalf("got %v, want %v", got, want)
		}
	}
	for _, intent := range ConfidentialIntents {
		if !envelopeIntentTypes[intent] {
			t.Fatalf("%q is not an allocated intent", intent)
		}
	}
}

func TestIntentRequiresEncryptionIsExact(t *testing.T) {
	for intent := range envelopeIntentTypes {
		if IntentRequiresEncryption(intent) != confidentialIntentSet[intent] {
			t.Fatalf("%q disagrees with the set", intent)
		}
	}
	for _, s := range []string{"Schedule_Meeting", " schedule_meeting", "schedule_meeting_response", "context_share2", "", "telepathy"} {
		if IntentRequiresEncryption(s) {
			t.Fatalf("%q matched", s)
		}
	}
}

func TestCheckEncryptionRequired(t *testing.T) {
	for _, intent := range ConfidentialIntents {
		r := CheckEncryptionRequired(map[string]interface{}{"intent": intent})
		if r.Allowed || r.Reason != "encryption_required" || r.Intent != intent {
			t.Fatalf("%s: got %+v", intent, r)
		}
	}
	for intent := range envelopeIntentTypes {
		if confidentialIntentSet[intent] {
			continue
		}
		if r := CheckEncryptionRequired(map[string]interface{}{"intent": intent}); !r.Allowed || r.Reason != "" || r.Intent != "" {
			t.Fatalf("%s: got %+v", intent, r)
		}
	}
	for name, env := range map[string]map[string]interface{}{
		"nil":               nil,
		"empty":             {},
		"non-string intent": {"intent": 7},
		"null intent":       {"intent": nil},
	} {
		if r := CheckEncryptionRequired(env); !r.Allowed {
			t.Fatalf("%s: got %+v", name, r)
		}
	}
	r := CheckEncryptionRequired(map[string]interface{}{"intent": "opportunity"}, "opportunity")
	if r.Allowed || r.Intent != "opportunity" {
		t.Fatalf("widened set: got %+v", r)
	}
	if r := CheckEncryptionRequired(map[string]interface{}{"intent": "ping"}, "opportunity"); !r.Allowed {
		t.Fatalf("widened set leaked: got %+v", r)
	}
}

File: test/encryption-policy.test.ts
import { describe, it, expect } from "vitest";
import {
  CONFIDENTIAL_INTENTS,
  intentRequiresEncryption,
  checkEncryptionRequired,
} from "../src/ink/encryption-policy.js";
import { IntentTypeSchema } from "../src/models/intent.js";

describe("CONFIDENTIAL_INTENTS", () => {
  it("names exactly the §3.4 set, every member an allocated intent", () => {
    expect([...CONFIDENTIAL_INTENTS].sort()).toEqual(["context_share", "multi_party_sync", "schedule_meeting"]);
    for (const intent of CONFIDENTIAL_INTENTS) {
      expect(IntentTypeSchema.safeParse(intent).success).toBe(true);
    }
  });

  it("intentRequiresEncryption agrees with the set", () => {
    for (const intent of IntentTypeSchema.options) {
      expect(intentRequiresEncryption(intent)).toBe((CONFIDENTIAL_INTENTS as readonly string[]).includes(intent));
    }
    expect(intentRequiresEncryption("telepathy")).toBe(false);
    expect(intentRequiresEncryption("")).toBe(false);
  });
});

describe("checkEncryptionRequired", () => {
  it("refuses every confidential intent in plaintext with encryption_required", () => {
    for (const intent of CONFIDENTIAL_INTENTS) {
      expect(checkEncryptionRequired({ intent })).toEqual({ allowed: false, reason: "encryption_required", intent });
    }
  });

  it("allows every other allocated intent", () => {
    for (const intent of IntentTypeSchema.options) {
      if ((CONFIDENTIAL_INTENTS as readonly string[]).includes(intent)) continue;
      expect(checkEncryptionRequired({ intent })).toEqual({ allowed: true });
    }
  });

  it("does not match by prefix, case or surrounding whitespace", () => {
    for (const intent of ["Schedule_Meeting", " schedule_meeting", "schedule_meeting_response", "context_share2"]) {
      expect(checkEncryptionRequired({ intent })).toEqual({ allowed: true });
    }
  });

  it("passes an envelope with no string intent to the schema, not the gate", () => {
    expect(checkEncryptionRequired({})).toEqual({ allowed: true });
    expect(checkEncryptionRequired({ intent: 7 })).toEqual({ allowed: true });
    expect(checkEncryptionRequired({ intent: null })).toEqual({ allowed: true });
    expect(checkEncryptionRequired(null)).toEqual({ allowed: true });
    expect(checkEncryptionRequired(undefined)).toEqual({ allowed: true });
  });

  it("lets a receiver widen the set with intents of its own", () => {
    const opts = { extraConfidentialIntents: ["opportunity"] };
    expect(checkEncryptionRequired({ intent: "opportunity" }, opts)).toEqual({
      allowed: false,
      reason: "encryption_required",
      intent: "opportunity",
    });
    expect(checkEncryptionRequired({ intent: "ping" }, opts)).toEqual({ allowed: true });
  });

  it("never lets a receiver narrow the protocol set", () => {
    const opts = { extraConfidentialIntents: ["opportunity"] };
    for (const intent of CONFIDENTIAL_INTENTS) {
      expect(checkEncryptionRequired({ intent }, opts)).toEqual({ allowed: false, reason: "encryption_required", intent });
      expect(checkEncryptionRequired({ intent }, { extraConfidentialIntents: [] })).toEqual({
        allowed: false,
        reason: "encryption_required",
        intent,
      });
    }
  });
});
